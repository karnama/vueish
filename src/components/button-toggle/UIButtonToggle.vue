<template>
    <UIButtonGroup :vertical="vertical" :spread="spread">
        <UIButton v-for="(option, index) in options"
                  :key="option.label"
                  :aria-checked="checked = isChecked(option)"
                  :category="secondary ? 'default' : 'primary'"
                  :outline="!checked"
                  :large="large"
                  v-bind="$attrs"
                  :disabled="disabled"
                  class="border"
                  :class="{
                      'border-gray-600': secondary,
                      'hover:border-gray-400 hover:bg-gray-400 hover:text-gray-900': secondary && !disabled,
                      'border-blue-600': !secondary,
                      'hover:border-blue-400': !secondary && !disabled && !checked,
                      'hover:border-blue-700': !secondary && !disabled && checked,
                      'text-blue-600': !checked && !secondary,
                      'hover:white-text': !checked && !secondary && !disabled,
                      'white-text bg-gray-600 disabled:bg-gray-600': checked && secondary,
                      'border-l-0': index > 0
                  }"
                  role="switch"
                  @click="toggleOption(option)"
                  @keydown.space="toggleOption(option)">
            <slot v-if="option.slot" :name="option.slot" :option="option" />
            <template v-else>
                {{ option.label }}
            </template>
        </UIButton>
    </UIButtonGroup>
</template>

<script lang="ts">
import { defineComponent, watch } from 'vue';
import type { PropType } from 'vue';
import UIButtonGroup, { props } from '@components/button-group/UIButtonGroup.vue';
import { large } from '@composables/style';
import { useVModel, disabled, clearable } from '@composables/input';
import UIButton from '@components/button/UIButton.vue';
import { isEqual, uniq } from 'lodash-es';
import { MaybeArray } from '@/types';
import { wrap } from '@/helpers';

export type Option = {
    label?: string; // label to use if no slot set
    value: any;
    slot?: string; // name of the slot to use for this option
};

export default defineComponent({
    name: 'UIButtonToggle',

    components: { UIButton, UIButtonGroup },

    inheritAttrs: false,

    props: {
        modelValue: {
            type: [Array, Object, null] as PropType<MaybeArray<Option> | null>
        },

        /**
         * The available options.
         */
        options: {
            type: Array as PropType<Option[]>,
            required: true,
            validator: (opts: Option[]) => uniq(opts.map(option => option.value)).length === opts.length
        },

        /**
         * Use the secondary styles.
         */
        secondary: {
            type: Boolean,
            default: false
        },

        /**
         * Property indicating whether multiple values can be selected or not.
         */
        multi: {
            type: Boolean,
            default: false
        },

        clearable,
        disabled,
        large,
        ...props
    },

    emits: ['update:modelValue'],

    setup(props) {
        const model = useVModel<MaybeArray<Option> | null>(props);

        const toggleOption = (option: Option): void => {
            const values = Array.isArray(model.value)
                ? model.value
                : model.value ? [model.value] : [];

            const optionIndex = values.findIndex(opt => isEqual(opt, option));

            if (props.multi) {
                if (!props.clearable && optionIndex !== -1 && values.length === 1) return;

                optionIndex === -1
                    ? (model.value as Option[]).push(option)
                    : (model.value as Option[]).splice(optionIndex, 1);

                return;
            }

            if (optionIndex === -1) {
                model.value = option;
                return;
            }

            if (!props.clearable) return;

            model.value = null;
        };
        const isChecked = (option: Option): boolean => {
            if (!model.value) {
                return false;
            }

            const values = Array.isArray(model.value) ? model.value : [model.value];

            return values.findIndex(val => isEqual(option.value, val.value)) !== -1;
        };

        watch(
            () => props.multi,
            val => {
                if (val) {
                    model.value = wrap(model.value).filter(v => !!v);
                    return;
                }

                if (Array.isArray(model.value)) {
                    model.value = model.value.length !== 1 ? null : model.value[0];
                }
            },
            { immediate: true }
        );

        return {
            model,
            isChecked,
            toggleOption
        };
    }
});
</script>

<style scoped lang="scss">
@variants hover {
    // specificity hack
    .white-text {
        @apply text-white;
    }
}
</style>
