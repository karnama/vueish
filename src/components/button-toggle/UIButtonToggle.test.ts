import { mount } from '@vue/test-utils';
import UIButtonToggle from './UIButtonToggle.vue';
import type { Option } from '@/types';

const options: Record<string, any>[] = [
    { label: 'Places', value: 'places' },
    { label: 'List', value: 'list' },
    { label: 'Map', value: 'map' }
] as Option[];

describe('UIButtonToggle', () => {
    it('should be able to select options', async () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options,
                modelValue: null
            }
        });

        await wrapper.get('button').trigger('click');
        expect(wrapper.emitted()).toHaveProperty('update:modelValue');
        expect(wrapper.lastEventValue()).toStrictEqual([options[0]]);
    });

    it('should allow for custom names slots', () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options: options.map(o => {
                    o.slot = o.value;
                    return o;
                }),
                modelValue: null
            },
            slots: options.reduce((prev, next) => {
                prev[next.value] = 'slotted content';
                return prev;
            }, {})
        });

        const buttons = wrapper.findAll('button');

        expect(buttons).toHaveLength(options.length);

        buttons.forEach(button => {
            expect(button.text()).toBe('slotted content');
        });
    });

    it('should be able to select multiple options given the prop', async () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options,
                modelValue: null,
                multi: true
            }
        });

        const buttons = wrapper.findAll('button');

        await buttons[0]!.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([[options[0]]]);
        await buttons[1]!.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([[options[0], options[1]]]);
        await buttons[1]!.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([[options[0]]]);
    });

    it('should be disabled given the prop', async () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options,
                modelValue: null,
                disabled: true
            }
        });

        await wrapper.get('button').trigger('click');
        expect(wrapper.lastEventValue()).toBeUndefined();
    });

    // eslint-disable-next-line jest/no-disabled-tests
    it.skip('should not be clearable by default', async () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options,
                modelValue: null
            }
        });

        const button = wrapper.get('button');

        await button.trigger('click');
        await button.trigger('click'); // not emitting if already selected
        expect(wrapper.emitted('update:modelValue')).toHaveLength(1);

        console.log(wrapper.vm.model);
        await wrapper.setProps({ clearable: true });
        console.log(wrapper.vm.model);
        await button.trigger('click');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(2);
        expect(wrapper.lastEventValue()).toStrictEqual([null]);
    });

    it('should clear up to a single value on multi select', async () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options,
                modelValue: null,
                multi: true
            }
        });

        const buttons = wrapper.findAll('button');

        await buttons[0]!.trigger('click');
        await buttons[1]!.trigger('click');

        expect(wrapper.lastEventValue()).toStrictEqual([[options[0], options[1]]]);

        await buttons[0]!.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([[options[1]]]);
        await buttons[1]!.trigger('click');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(4);
    });

    it('should be clearable given the prop', async () => {
        const wrapper = mount(UIButtonToggle, {
            props: {
                options,
                modelValue: null,
                clearable: true
            }
        });

        const buttons = wrapper.findAll('button');
        await buttons[0]!.trigger('click');
        await buttons[0]!.trigger('click');

        expect(wrapper.lastEventValue()).toStrictEqual([null]);

        await wrapper.setProps({ multi: true });
        await buttons[0]!.trigger('click');
        await buttons[1]!.trigger('click');

        await buttons[0]!.trigger('click');
        await buttons[1]!.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([[]]);
    });
});
