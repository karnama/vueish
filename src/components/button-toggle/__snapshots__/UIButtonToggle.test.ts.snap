// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIButtonToggle should correctly display 1`] = `
<div
  class="button-group"
  role="group"
>
  
  
  <button
    aria-checked="false"
    class="text-gray-600 border shadow-sm border-gray-400 disabled:bg-gray-300 disabled:shadow-none hover:text-white hover:bg-gray-400 px-3.5 py-2 ui-button rounded relative font-bold text-sm m-0 border"
    role="switch"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label transition-opacity duration-200"
      >
        
        
        Places
        
        
      </span>
      
    </transition-stub>
  </button>
  <button
    aria-checked="false"
    class="text-gray-600 border shadow-sm border-gray-400 disabled:bg-gray-300 disabled:shadow-none hover:text-white hover:bg-gray-400 px-3.5 py-2 ui-button rounded relative font-bold text-sm m-0 border border-l-0"
    role="switch"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label transition-opacity duration-200"
      >
        
        
        List
        
        
      </span>
      
    </transition-stub>
  </button>
  <button
    aria-checked="false"
    class="text-gray-600 border shadow-sm border-gray-400 disabled:bg-gray-300 disabled:shadow-none hover:text-white hover:bg-gray-400 px-3.5 py-2 ui-button rounded relative font-bold text-sm m-0 border border-l-0"
    role="switch"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label transition-opacity duration-200"
      >
        
        
        Map
        
        
      </span>
      
    </transition-stub>
  </button>
  
  
</div>
`;
